shap/explainers/_explainer.py

```diff
@@ -117,6 +123,10 @@ def __init__(self, model, masker=None, link=links.identity, algorithm="auto", ou
                     models.TransformersPipeline(self.model), self.masker,
                     link=link, algorithm=algorithm, output_names=output_names, feature_names=feature_names, linearize_link=linearize_link, **kwargs
                 )
+        if masker_length is not None and hasattr(self.masker, 'max_samples') and masker_length > self.masker.max_samples:
+            logging.warn(f"Your background dataset is larger than the max_samples {self.masker.max_samples}. You can hand over max_samples by explicitly "
+                         "defining the masker to use: "
+                         "`shap.Explainer(model, masker=maskers.Independent(X, max_samples=1000))`")
```

Comment:
A few suggestions:

1. This seems possibly an odd place for the warning to be raised. Would it be more appropriate to raise it in the masker itself, where the subsampling actually occurs? I think this would be a bit tidier.

https://github.com/CloseChoice/shap/blob/b23058409eb05f5cb9e7e5cc7c59ea22b38a5695/shap/maskers/_tabular.py#L57-L58

2. It could help to adopt the common pattern of using a module-specific logger, e.g. `log = logging.getLogger(__name__)`

3. I believe `.warn` is deprecated in favour of `.warning`. 

4. I think the warning message could perhaps explain the consequences of having more samples, e.g.:
```python
log.warning(
    "The background dataset exceeds the masker's `max_samples`, "
    "so will be subsampled from {n} rows to {max_samples} rows."
)
```

Reply:
I agree that it's an odd place but with this approach we do not need to repeat code in the maskers itself but have one place where this check happens. Would you rather duplicate the code in the maskers (I believe we currently have 2 maskers with the `max_samples` keyword)?

As for the other points I totally agree and will implement these when we've settled the point above.

Reply:
Good point. On balance I'd probably prefer to log a message in both maskers. That way, the same warning is raised whether the old shap_values or the new Explainer interface is used.

One other thing actually, going back to the discussion on #3461 : I think there is a risk that excessive logging could be annoying for most users. We had originally proposed logging.INFO, but here is logging.WARNING. I think the latter will display in an IPython or Jupyter output by default.

Is this desirable? IMO the subsampling is pretty usual behaviour, and will be invoked probably the majority of the time, so it seems excessive to have a warning message visible on most runs of the package! I would think a visible warning would only show for things that are unusual / unintended / require user attention.

Reply:
to be honest, might be that our behaviour is standard but I still think it's bad practice to change user behaviour without at least an info (I would even prefer a warning and think it would be better if the user manually configures the correct `max_samples` if the provided data exceeds the default.)

Reply:
>  it's bad practice to change user behaviour without ...

Please would you elaborate on this point: does the code "change user behaviour"?

Reply:
I mean the user selects 1000 rows, potentially a well balanced dataset and we sample this down to 100 destroying the balance. Even if something like this is standard behaviour I wouldn't consider this good practice. At least we should inform the user how to use the balanced dataset WITHOUT sampling it down.

Reply:
How does one use background dataset without sampling it down? It seems no matter what I do the background dataset gets subsampled. 

When looking at the code, almost the very first lines in the code (starting line 166) are

        masker = data
        super().__init__(model, masker, feature_names=feature_names)

        if type(self.masker) is maskers.Independent:
            data = self.masker.data
        elif masker is not None:
            raise InvalidMaskerError(f"Unsupported masker type: {str(type(self.masker))}!")

The second line subsamples the 'masker' to 100 which then gets assigned to the data. This later on gets assigned to self.data (line 182) which then used to generate baseline on line 229

         self.expected_value = self.model.predict(self.data).mean(0)

Is this the intended behavior or a bug?

Reply:
No, it isn't. I am using the command "shap.TreeExplainer" instead of "shap.Explainer". Does it make a difference? I've tried many different thing and nothing seems to work. The line with "super()." seems to downsample background data no matter what you do. The default value of 100 for "max_sample" ends up being used every time.

Reply:
Please see https://github.com/shap/shap/issues/3680